python/binding/tritonserver_pybind.cc

```diff
@@ -0,0 +1,1991 @@
+};
+struct NotFound : public TritonError {
+  explicit NotFound(const std::string& what) : TritonError(what) {}
+};
+struct InvalidArgument : public TritonError {
+  explicit InvalidArgument(const std::string& what) : TritonError(what) {}
+};
+struct Unavailable : public TritonError {
+  explicit Unavailable(const std::string& what) : TritonError(what) {}
+};
+struct Unsupported : public TritonError {
+  explicit Unsupported(const std::string& what) : TritonError(what) {}
+};
+struct AlreadyExists : public TritonError {
+  explicit AlreadyExists(const std::string& what) : TritonError(what) {}
+};
+
+TRITONSERVER_Error*
+CreateTRITONSERVER_ErrorFrom(const py::error_already_set& ex)
+{
+  // Reserved lookup to get Python type of the exceptions,
+  // 'TRITONSERVER_ERROR_UNKNOWN' is the fallback error code.
+  // static auto uk = py::module::import("triton_bindings").attr("Unknown");
+  static auto it = py::module::import("triton_bindings").attr("Internal");
+  static auto nf = py::module::import("triton_bindings").attr("NotFound");
+  static auto ia =
+      py::module::import("triton_bindings").attr("InvalidArgument");
+  static auto ua = py::module::import("triton_bindings").attr("Unavailable");
+  static auto us = py::module::import("triton_bindings").attr("Unsupported");
+  static auto ae = py::module::import("triton_bindings").attr("AlreadyExists");
+  TRITONSERVER_Error_Code code = TRITONSERVER_ERROR_UNKNOWN;
+  if (ex.matches(it.ptr())) {
+    code = TRITONSERVER_ERROR_INTERNAL;
+  } else if (ex.matches(nf.ptr())) {
+    code = TRITONSERVER_ERROR_NOT_FOUND;
+  } else if (ex.matches(ia.ptr())) {
+    code = TRITONSERVER_ERROR_INVALID_ARG;
+  } else if (ex.matches(ua.ptr())) {
+    code = TRITONSERVER_ERROR_UNAVAILABLE;
+  } else if (ex.matches(us.ptr())) {
+    code = TRITONSERVER_ERROR_UNSUPPORTED;
+  } else if (ex.matches(ae.ptr())) {
+    code = TRITONSERVER_ERROR_ALREADY_EXISTS;
+  }
+  return TRITONSERVER_ErrorNew(code, ex.what());
+}
+
+void
+ThrowIfError(TRITONSERVER_Error* err)
+{
+  if (err == nullptr) {
+    return;
+  }
+  std::shared_ptr<TRITONSERVER_Error> managed_err(
+      err, TRITONSERVER_ErrorDelete);
+  std::string msg = TRITONSERVER_ErrorMessage(err);
+  switch (TRITONSERVER_ErrorCode(err)) {
+    case TRITONSERVER_ERROR_INTERNAL:
+      throw Internal(std::move(msg));
+    case TRITONSERVER_ERROR_NOT_FOUND:
+      throw NotFound(std::move(msg));
+    case TRITONSERVER_ERROR_INVALID_ARG:
+      throw InvalidArgument(std::move(msg));
+    case TRITONSERVER_ERROR_UNAVAILABLE:
+      throw Unavailable(std::move(msg));
+    case TRITONSERVER_ERROR_UNSUPPORTED:
+      throw Unsupported(std::move(msg));
+    case TRITONSERVER_ERROR_ALREADY_EXISTS:
+      throw AlreadyExists(std::move(msg));
+    default:
+      throw Unknown(std::move(msg));
+  }
+}
+
+template <typename TritonStruct>
+class PyWrapper {
+ public:
+  explicit PyWrapper(TritonStruct* to, bool o) : triton_object_(to), owned_(o)
```

Comment:
recommend: consistent naming for 'owned' - some places below we use 'owned'

same for triton_object